torch/csrc/jit/passes/bailout_graph.cpp

```diff
@@ -0,0 +1,203 @@
+      auto block_outputs = n->owningBlock()->outputs();
+      // skip the first input for loops (current iteration count)
+      size_t i = outer_node->kind() == prim::Loop;
+      auto new_outputs = outer_node->kind() == prim::Loop
+          ? outer_node->inputs()
+          : outer_node->outputs();
+      for (; i < block_outputs.size(); i++) {
+        auto nv = old_to_new_[block_outputs[i]];
+        old_to_new_[new_outputs.at(i)] = nv;
+      }
+      buildBailOutBlockFrom(outer_node);
+    }
+  }
+
+  std::shared_ptr<Graph> buildBailOutGraphFrom(Node* n) {
+    old_to_new_.clear();
+    copy_graph_ = std::make_shared<Graph>();
+    buildBailOutBlockFrom(n);
+    // add graph outputs
+    for (auto ov : graph_->outputs()) {
+      auto new_value =
+          (this->old_to_new_.count(ov) == 0) ? nullptr : this->old_to_new_[ov];
+      auto nv = addNewInputForValue(ov, new_value);
+      copy_graph_->registerOutput(nv);
+    }
+    return copy_graph_;
+  }
+
+  std::shared_ptr<Graph> graph_;
+  std::shared_ptr<Graph> copy_graph_;
+  std::vector<Value*> live_inputs_;
+  std::unordered_map<Value*, Value*> old_to_new_;
+};
+
+struct BailOutInserter {
+  BailOutInserter(std::shared_ptr<Graph> graph) : graph_(std::move(graph)) {}
+
+  void run() {
+    insertBailOuts(graph_->block());
+    // set bailout subgraphs
+    setSubgraphs();
+  }
+
+  // remove all BailOuts and Guards
+  // we will profile this graph
+  // and insert guards and bailouts again
+  // where needed if its guard fails
+  void sanitizeGraph(Block* b) {
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); ++it) {
+      if (it->kind() == prim::Guard) {
+        // this will need to be profiled again
+        it->input()->setType(TensorType::create());
+        // destroy the guard
+        it->output()->replaceAllUsesWith(it->input());
+        it.destroyCurrent();
+      }
+
+      if (it->kind() == prim::BailOut) {
+        // this is the index of an input we are actually guarding
+        auto index = it->i(attr::slot);
+        // this will need to be profiled again
+        it->inputs()[index]->setType(TensorType::create());
+        // destroy the guard
+        it->output()->replaceAllUsesWith(it->inputs()[index]);
+        it.destroyCurrent();
+      }
+
+      for (auto ib : it->blocks()) {
+        sanitizeGraph(ib);
+      }
+    }
+  }
+
+  void setSubgraphs() {
+    for (auto e : subgraphs) {
+      sanitizeGraph(e.second->block());
+      ConstantPooling(e.second);
+      e.first->g_(attr::Subgraph, e.second);
+    }
+  }
+
+  void insertBailOuts(Block* b) {
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); ++it) {
+      if (it->kind() == prim::Guard) {
+        auto bailout_node = b->owningGraph()->create(prim::BailOut);
+        auto node = *it;
+
+        BailOutGraphBuilderForNode bg(graph_);
+        auto bailout_graph = bg.buildBailOutGraphFrom(node);
+
+        // N.B. we don't immediately set SubGraph attribute
```

Comment:
This comment isn't very clear. Generally there are not enough comments in this pass for me to understand the structure.

Reply:
From more reading, it seems like this is because we editing in BailOut nodes, mutating the graph while we are also trying to generate continuations. This is already doing half of the work later editing the graph? Why not do _all_ of the editing later? Then it would not be necessary to delete BailOut nodes from continuations as well.